source/base/mpi_remote_point_evaluation.cc

```diff
@@ -0,0 +1,448 @@
+
+      indices_ptr = {0}; // TODO: name
+
+      for (const auto &i : relevant_points_per_process_offset)
+        {
+          const unsigned int rank = i.first;
+
+
+          const auto &relevant_points_offset =
+            relevant_points_per_process_offset[rank];
+          const auto &relevant_points_count =
+            relevant_points_per_process_count[rank];
+
+          unsigned int c = 0;
+
+          for (unsigned int j = 0; j < relevant_points_offset.size(); ++j)
+            for (unsigned int k = 0; k < relevant_points_count[j]; ++k, ++c)
+              {
+                auto &qp_counter =
+                  quadrature_points_count[relevant_points_offset[j]];
+                indices_temp.emplace_back(relevant_points_offset[j],
+                                          qp_counter);
+
+                ++qp_counter;
+              }
+
+          indices_ptr.push_back(indices_ptr.back() + c);
+        }
+
+      quadrature_points_ptr = {0};
+
+      this->unique_mapping = true;
+      for (const auto &i : quadrature_points_count)
+        {
+          this->unique_mapping &= (i == 1);
+          quadrature_points_ptr.push_back(quadrature_points_ptr.back() + i);
+        }
+
+      indices = {};
+      for (const auto &i : indices_temp)
+        indices.push_back(quadrature_points_ptr[i.first] + i.second);
+
+      this->relevant_remote_points_per_process = {};
+
+      send_ranks = {};
+      send_ptr   = {0};
+
+      for (const auto &i : relevant_remote_points_per_process)
+        {
+          auto &temp = this->relevant_remote_points_per_process;
+
+          std::vector<std::tuple<std::pair<int, int>, unsigned int, Point<dim>>>
+            full;
+          full.reserve(i.second.size());
+
+          for (const auto &j : i.second)
+            full.emplace_back(j.first, full.size(), j.second);
+
+          std::sort(full.begin(), full.end(), [](const auto &a, const auto &b) {
+            if (std::get<0>(a) != std::get<0>(b))
+              return std::get<0>(a) < std::get<0>(b);
+            else
+              return std::get<1>(a) < std::get<1>(b);
+          });
+
+          std::pair<int, int> current = {-1, -1};
+
+          for (const auto &j : full)
+            {
+              if (current != std::get<0>(j))
+                {
+                  current = std::get<0>(j);
+                  std::get<0>(temp).emplace_back(current, 0);
+                }
+
+              std::get<0>(temp).back().second++;
+
+              std::get<1>(temp).push_back(std::get<2>(j));
+              std::get<2>(temp).push_back(std::get<1>(j) + send_ptr.back());
+            }
+
+          send_ranks.push_back(i.first);
+          send_ptr.push_back(std::get<2>(temp).size());
+        }
+#endif
+    }
+
+
+    template <int dim, int spacedim>
+    const std::vector<unsigned int> &
+    RemotePointEvaluation<dim, spacedim>::get_point_ptrs() const
+    {
+      return quadrature_points_ptr;
+    }
```

Comment:
I think this function (and the small ones below) can go to the header file.

Reply:
Yes. I can do that. What is the criterion when to leave a function in the header file. I would have thought that there are only two reasons: 1) that the function is templated in such a way that explicit instantiation is hard and 2) the calls is performance critical (as is in many places of `FEEvaluation`). But the functions here are definitely not performance critical so that they do not need to be inlined...